compiler/src/dotty/tools/dotc/reporting/MessageRendering.scala

```diff
@@ -166,35 +214,64 @@ trait MessageRendering {
   /** The whole message rendered from `msg` */
   def messageAndPos(dia: Diagnostic)(using Context): String = {
     import dia._
-    val levelString = diagnosticLevel(dia)
+    val pos1 = pos.nonInlined
+    val inlineStack = inlinePosStack(pos).filter(_ != pos1)
+    val maxLineNumber =
+      if pos.exists then (pos1 :: inlineStack).map(_.endLine).max + 1
+      else 0
+    given Level = Level(level)
+    given Offset = Offset(maxLineNumber.toString.length + 2)
     val sb = mutable.StringBuilder()
-    val posString = posStr(pos, levelString, msg)
+    val posString = posStr(pos, msg, diagnosticLevel(dia))
     if (posString.nonEmpty) sb.append(posString).append(EOL)
     if (pos.exists) {
       val pos1 = pos.nonInlined
       if (pos1.exists && pos1.source.file.exists) {
-        val (srcBefore, srcAfter, offset) = sourceLines(pos1, levelString)
-        val marker = columnMarker(pos1, offset, levelString)
-        val err = errorMsg(pos1, msg.message, offset)
-        sb.append((srcBefore ::: marker :: err :: outer(pos, " " * (offset - 1)) ::: srcAfter).mkString(EOL))
+        val (srcBefore, srcAfter, offset) = sourceLines(pos1)
+        val marker = positionMarker(pos1)
+        val err = errorMsg(pos1, msg.message)
+        sb.append((srcBefore ::: marker :: err :: srcAfter).mkString(EOL))
+
+        if inlineStack.nonEmpty then
+          sb.append(EOL).append(newBox())
+          sb.append(EOL).append(offsetBox).append(i"Inline stack trace")
+          for inlinedPos <- inlineStack if inlinedPos != pos1 do
+            val (srcBefore, srcAfter, offset) = sourceLines(inlinedPos)
+            val marker = positionMarker(inlinedPos)
+            sb.append(EOL).append(newBox(soft = true))
+            sb.append(EOL).append(offsetBox).append(i"This location contains code that was inlined from $pos")
+            sb.append(EOL).append((srcBefore ::: marker :: srcAfter).mkString(EOL))
+          sb.append(EOL).append(endBox)
       }
       else sb.append(msg.message)
     }
     else sb.append(msg.message)
     if (dia.isVerbose)
       appendFilterHelp(dia, sb)
+
+    if Diagnostic.shouldExplain(dia) then
+      sb.append(EOL).append(newBox())
+      sb.append(EOL).append(offsetBox).append(" Explanation")
```

Comment:
To make it clear that this comes from -explain (since people might not realize it's enabled in their build):
```suggestion
      sb.append(EOL).append(offsetBox).append(" Explanation (enabled by `-explain`)")
```